### PR TITLE
Add scripts to run all tests in exercism/elm repo

### PR DESCRIPTION
**Pull request by jiegillet**
With these scripts, it will be possible to run `elm-test-runner` on each exercise from the exercism/elm repo in the CI, as discussed in [this issue](https://github.com/exercism/elm/issues/412).

To test locally `bin/run-all-exercises-in-docker.sh path/to/elm/repo`.

When I try it, it's very slow and I get a bunch of "tinyproxy: Could not create listening sockets." so maybe the tinyproxy script needs to be modified, but I would like to see the behavior in the CI here before touching it.

### CODE REVIEW

**Comment by mpizenberg** [comment]
I can't check today but here are couple infos I remember.

First, Tiny proxy is used in the container to create a proxy server that
automatically fail all http requests. This is the only I found to prevent
the elm compiler to wait for roughly 5 seconds doing nothing when on Linux
without network.

From memory, we do not check whether the proxy port is available, just
assume it is. So I think I already had that tiny proxy error message in a
context where I try to run it multiple times, resulting in the first time
succeeding and all subsequent ones to fail due to occupied port. So the
error might not be problematic.

Then regarding the long runtime duration. I cannot say exactly before
'ooking at it. One thing for sure is that due to the cost of starting Node,
it is much more expensive to run tests one after each other, than to craft
an executable concatenating all tests as done in CI of our main repo.

If you're running independant docker calls per test run. The cost per run
is probably around 1s.

**Comment by jiegillet** [comment]
Mmh... There are still "tinyproxy: Could not create listening sockets." messages in the smoke-tests, but they are still very fast, so I guess it's OK? Maybe the proxy isn't required when we run docker with `--network none`?

This is one of those times docker frustrates me because its whole job is to make behavior reproducible.

**Comment by jiegillet** [comment]
> So I think I already had that tiny proxy error message in a context where I try to run it multiple times, resulting in the first time succeeding and all subsequent ones to fail due to occupied port. So the error might not be problematic. Then regarding the long runtime duration. I cannot say exactly before looking at it. 

Ah you are right, it's probably the repeated use. Not a problem then.

> One thing for sure is that due to the cost of starting Node, it is much more expensive to run tests one after each other, than to craft an executable concatenating all tests as done in CI of our main repo. If you're running independant docker calls per test run. The cost per run is probably around 1s.

Yes, I considered that, it makes sense when you just want to run the tests, but here we also want to make sure that the test code is extracted and parsed without issues (like we had that one time when tests had identical names). To do that, we have no choice but to run elm-test-runner on each folder.

I still don't know why it runs fast here and slow on my machine, but this is much less problematic than the opposite situation :)

**Comment by jiegillet** [comment]
@mpizenberg it would be great to have your review for this one

**Review by mpizenberg — APPROVED** [review]
This seems to be working well. I've checked that I get no error when all tests pass and an error when some test do not pass.
I've also looked if we could easily make the test run parallel but it didn't work.
The trick would be to put the last for loop content in a bash function, and do the loop with a function call and a `&` at the end to fork in background.
Unfortunately, the inside of the `bin/run.sh` script is not super parallel-friendly with files and all so it crashed and I didn't spend a lot of time to try figuring it out.

For me, the sequence runs in about 48s, with approximately 1s per exercise.

**Comment by jiegillet** [comment]
Thank you for the review and QA!

I am thinking now about an exercise that might require two solution files, in this case the script will break, so I will have to revisit it, maybe I can take the opportunity to look what I could do to make `bin/run.sh` more parallel friendly. 
I'll open an issue to remember this.

@exercism/maintainers-admin could I get an approval for this?

File: Dockerfile
Comment (ErikSchierboom) [inline Dockerfile:45]: Instead of including the shell file into the docker image (which is fine), an alternative is to map the file into the container at runtime: https://github.com/exercism/generic-test-runner/blob/main/bin/run-tests-in-docker.sh#L28

**Review by ErikSchierboom — APPROVED** [review]

File: Dockerfile
Comment (jiegillet) [reply Dockerfile:45]: Oh that means I could be running this entirely from the elm repo, that might be a better choice. I will try it out before merging this. 
Thank you!

Comment (jiegillet) [reply Dockerfile:45]: OK cool, it works, I made it happen [in this PR](https://github.com/exercism/elm/pull/527).
I guess there are pros and cons to having it in the elm repo.

Pros:
- the script lives in the same repo than the CI running it
- easier to modify the script when we change assumptions on how the elm exercises are structured

Cons:
- the script makes assumptions on how the test-runner is structured
- it's a script about running tests, this repo is a better fit

In the end, there are assumptions on the structure in both directions, but I feel like the test-runner is not going to change as much, so I'm leaning towards having it in the elm repo.
@mpizenberg, what do you think?

Comment (mpizenberg) [reply Dockerfile:45]: agreed, this matters mainly for the elm repo, as this is the one we wanted to have these tests for. So let's keep it in the main elm repo

**Comment by jiegillet** [comment]
Closed in favor of https://github.com/exercism/elm/pull/527

**Pull request closed without merging**
